runner/gazelle.go

```diff
@@ -0,0 +1,167 @@
+/* Copyright 2016 The Bazel Authors. All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+   http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// Command gazelle is a BUILD file generator for Go projects.
+// See "gazelle --help" for more details.
+package runner
+
+import (
+	"flag"
+	"fmt"
+	"log"
+	"os"
+
+	"github.com/bazelbuild/bazel-gazelle/config"
+	"github.com/bazelbuild/bazel-gazelle/language"
+)
+
+type command int
+
+const (
+	updateCmd command = iota
+	fixCmd
+	updateReposCmd
+	helpCmd
+)
+
+var commandFromName = map[string]command{
+	"fix":          fixCmd,
+	"help":         helpCmd,
+	"update":       updateCmd,
+	"update-repos": updateReposCmd,
+}
+
+var nameFromCommand = []string{
+	// keep in sync with definition above
+	"update",
+	"fix",
+	"update-repos",
+	"help",
+}
+
+func (cmd command) String() string {
+	return nameFromCommand[cmd]
+}
+
+func Main(langs []language.Language) {
+	log.SetPrefix("gazelle: ")
+	log.SetFlags(0) // don't print timestamps
+
+	var wd string
+	if wsDir := os.Getenv("BUILD_WORKSPACE_DIRECTORY"); wsDir != "" {
+		wd = wsDir
+	} else {
+		var err error
+		if wd, err = os.Getwd(); err != nil {
+			log.Fatal(err)
+		}
+	}
+
+	if err := Run(langs, wd, os.Args[1:]); err != nil && err != flag.ErrHelp {
+		if err == ErrExit {
+			os.Exit(1)
+		} else {
+			log.Fatal(err)
+		}
+	}
+}
+
+func Run(langs []language.Language, wd string, args []string) error {
```

Comment:
Can add docs!  My thinking was that most callers would just call `Main([]langs)` and be done with it; but in certain situations the internal `os.Exit()` would be a problem, like if you're plugging gazelle into another process.

Reply:
I would suggest to move `Main()` back into `cmd/gazelle` and just expose `Run()` for now.
In the doc for `Run()` you could reference to Main and/or provide an example usage in runner/gazelle_test.go (see https://pkg.go.dev/testing#hdr-Examples)

In the future if there is a need for `Main` then we can move it out later?

Reply:
Ah interesting.  I figured most callers (including myself) would mostly only care about Main(), and exposing Run() is the nice-to-have.  In fact, I was just considering reverting the export of `ErrExit` since I'm finding it difficult to document accurately.  If I move Main() back out to gazelle then I have to export `ErrExit`.